### PR TITLE
DEV: Update modifyclass docs

### PR DESCRIPTION
**Pull request by davidtaylorhq**
Previously this topic only covered a single edge case. This commit updates it to be a general introduction to modifyClass, and moves the edge-case to a new 'troubleshooting' section

Recommend reviewing [with whitespace changes hidden](https://github.com/discourse/discourse-developer-docs/pull/3/files?diff=split&w=1)

### CODE REVIEW

File: docs/03-code-internals/14-modifyclass.md [hunk 1]
```diff
@@ -1,71 +1,138 @@
...
+      @action
+      click() {
+        console.log("button was clicked");
+        super.click();
+      }
+    }
+);
 ```
 
+The `class extends ...` syntax mimics that of [JS child classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#inheritance). In general, any syntax/features supported by child classes can be applied here. That includes `super`, static properties/functions, and more.
```
Comment (janzenisaac) [inline docs/03-code-internals/14-modifyclass.md:41]: We should include documentation on the side affects of omitting `super` when extending a class

File: docs/03-code-internals/14-modifyclass.md [hunk 2]
```diff
@@ -1,71 +1,138 @@
...
 
-### Adding a `lookup()` caused the error
+## Basic Usage
 
-If you `lookup()` a singleton too early in the boot process, it will cause any later `modifyClass` calls to fail. In this situation, you should try to move the lookup to happen later. For example, you would change something like this:
+`api.modifyClass` can be used to modify the functions and properties of any class which is accessible via the Ember resolver. That includes Discourse's routes, controllers, services and components.
 
+`modifyClass` takes two arguments:
+
+- `resolverName` (string) - construct this by using the type (e.g. component/controller/etc.), followed by a colon, followed by the (dasherized) filename name of the class. For example: `component:d-button`, `controller:user`, `route:application`, etc.
```
Comment (janzenisaac) [inline docs/03-code-internals/14-modifyclass.md:21]: We can touch on nested components here as well.

Comment (davidtaylorhq) [reply docs/03-code-internals/14-modifyclass.md:21]: Added a nested example :+1:

File: docs/03-code-internals/14-modifyclass.md (hunk 1 above)
Comment (davidtaylorhq) [reply docs/03-code-internals/14-modifyclass.md:41]: Can you expand on this? I'm not sure it's clear-cut - it really depends what you're doing. e.g. if you're extending a method like `didInsertElement()`, then you almost certainly want to call the core implementation first. But if you're aiming to change behaviour (e.g. overriding a `click()` action) then calling the core implementation might not make sense.

File: docs/03-code-internals/14-modifyclass.md [hunk 3]
```diff
@@ -1,71 +1,138 @@
...
+The `class extends ...` syntax mimics that of [JS child classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#inheritance). In general, any syntax/features supported by child classes can be applied here. That includes `super`, static properties/functions, and more.
 
-To this:
+However, there are some limitations. The `modifyClass` system only detects changes to the class's JS `prototype`. Practically, that means:
+
+- introducing or modifying a `constructor()` is not supported
+
+- introducing or modifying class fields is not supported (although some decorated class fields, like `@tracked` can be used)
+
+- simple class fields on the original implementation cannot be overridden in any way (although, as above, `@tracked` fields can be overriden by another `@tracked` field)
```
[the remark this answers is not in the document]
Comment (davidtaylorhq) [reply docs/03-code-internals/14-modifyclass.md]: Added

File: docs/03-code-internals/14-modifyclass.md [hunk 4]
```diff
@@ -1,71 +1,169 @@
 ---
-title: Using modifyClass for objects which are initialized early in boot
-short_title: ModifyClass
+title: Using modifyClass to change core behaviour
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:2]: 🦅 🍔 🎆 😄 

```suggestion
title: Using modifyClass to change core behavior
```

File: docs/03-code-internals/14-modifyclass.md [hunk 5]
```diff
@@ -1,71 +1,169 @@
...
 ---
-When using modifyClass in an initializer, you may see this warning in the console:
 
-> Attempted to modify "{name}", but it was already initialized earlier in the boot process (e.g. via a lookup()). Remove that lookup, or move the modifyClass call earlier in the boot process for changes to take effect.
+For advanced themes and plugins, Discourse offers the `modifyClass` system. This allows you to extend and override functionality in many of core's javascript classes.
 
-In theme/plugin development, there are two ways this error is normally introduced:
+## When to use `modifyClass`
+
+`modifyClass` should be a last resort, when your customization cannot be made via Discourse's more stable customization APIs (e.g. plugin-api methods, plugin outets, transformers).
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:11]:
```suggestion
`modifyClass` should be a last resort, when your customization cannot be made via Discourse's more stable customization APIs (e.g. plugin-api methods, plugin outlets, transformers).
```

File: docs/03-code-internals/14-modifyclass.md [hunk 6]
```diff
@@ -1,71 +1,169 @@
...
+- introducing or modifying class fields is not supported (although some decorated class fields, like `@tracked` can be used)
+
+  ```js
+  api.modifyClass("component:foo", (Superclass) => class extends Superclass {
+    someField = "foo"; // NOT SUPPORTED - do not copy
+    @tracked someOtherField = "foo"; // This is ok
+  })
+  ```
+
+- simple class fields on the original implementation cannot be overridden in any way (although, as above, `@tracked` fields can be overriden by another `@tracked` field)
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:64]:
```suggestion
- simple class fields on the original implementation cannot be overridden in any way (although, as above, `@tracked` fields can be overridden by another `@tracked` field)
```

File: docs/03-code-internals/14-modifyclass.md [hunk 7]
```diff
@@ -1,71 +1,169 @@
...
+  // or
+  // (theme)/javascripts/discourse/pre-initializers/extend-user-for-my-plugin.js
+
+  import { withPluginApi } from "discourse/lib/plugin-api";
+
+  export default {
+    name: "extend-user-for-my-plugin",
+    before: "inject-discourse-objects",
+
+    initializeWithApi(api){
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:155]:
```suggestion
    initializeWithApi(api) {
```

File: docs/03-code-internals/14-modifyclass.md [hunk 8]
```diff
@@ -1,71 +1,169 @@
...
-// 'Just in time' lookup of service (good!)
-export default apiInitializer("0.8", (api) => {
-  api.composerBeforeSave(async () => {
-    const composerService = api.container.lookup("service:composer");
-    composerService.doSomething();
-  });
-});
+// Outdated syntax - do not use
+api.modifyClass("component:some-component", {
+  someFunction(){
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:87]:
```suggestion
  someFunction() {
```

File: docs/03-code-internals/14-modifyclass.md [hunk 9]
```diff
@@ -1,71 +1,169 @@
...
+    someField = "foo"; // NOT SUPPORTED - do not copy
+    @tracked someOtherField = "foo"; // This is ok
+  })
+  ```
+
+- simple class fields on the original implementation cannot be overridden in any way (although, as above, `@tracked` fields can be overriden by another `@tracked` field)
+
+  ```js
+  // Core code:
+  class Foo extends Component{
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:68]:
```suggestion
  class Foo extends Component {
```

File: docs/03-code-internals/14-modifyclass.md [hunk 10]
```diff
@@ -1,71 +1,169 @@
...
+The `class extends ...` syntax mimics that of [JS child classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes#inheritance). In general, any syntax/features supported by child classes can be applied here. That includes `super`, static properties/functions, and more.
+
+However, there are some limitations. The `modifyClass` system only detects changes to the class's JS `prototype`. Practically, that means:
 
-To this:
+- introducing or modifying a `constructor()` is not supported
+
+  ```js
+  api.modifyClass("component:foo", (Superclass) => class extends Superclass {
+    constructor(){
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:49]:
```suggestion
    constructor() {
```

File: docs/03-code-internals/14-modifyclass.md [hunk 11]
```diff
@@ -1,71 +1,169 @@
...
-  });
-});
+// Outdated syntax - do not use
+api.modifyClass("component:some-component", {
+  someFunction(){
+    const original = this._super();
+    return original + " some change";
+  }
+  pluginId: "some-unique-id"
+})
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:92]:
```suggestion
});
```

File: docs/03-code-internals/14-modifyclass.md [hunk 12]
```diff
@@ -1,71 +1,169 @@
...
+  })
+  ```
+
+- introducing or modifying class fields is not supported (although some decorated class fields, like `@tracked` can be used)
+
+  ```js
+  api.modifyClass("component:foo", (Superclass) => class extends Superclass {
+    someField = "foo"; // NOT SUPPORTED - do not copy
+    @tracked someOtherField = "foo"; // This is ok
+  })
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:61]:
```suggestion
  });
```

File: docs/03-code-internals/14-modifyclass.md [hunk 13]
```diff
@@ -1,71 +1,169 @@
...
 
-To this:
+- introducing or modifying a `constructor()` is not supported
+
+  ```js
+  api.modifyClass("component:foo", (Superclass) => class extends Superclass {
+    constructor(){
+      // This is not supported. The constructor will be ignored
+    }
+  })
```
Comment (CvX) [inline docs/03-code-internals/14-modifyclass.md:52]:
```suggestion
  });
```

**Review by CvX — APPROVED** [review]
just minor typos and consistency notes 😃✅

**Pull request merged**